File: backend/app/crawlers/bilibili_crawler.py
import httpx
import random
from typing import List, Dict, Any

class BilibiliCrawler:
    """Crawler Bilibili bất đồng bộ - Lọc sạch rác quảng cáo & bvid rỗng, chuẩn 16 video sạch"""
    
    BASE_SEARCH_URL = "https://api.bilibili.com/x/web-interface/wbi/search/type"
    BASE_HOME_URL = "https://api.bilibili.com/x/web-interface/popular"

    HEADERS = {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
        "Referer": "https://www.bilibili.com/"
    }

    # Danh sách từ khóa hot xoay vòng để lấy data gợi ý tươi mới liên tục
    DYNAMIC_TOPICS = ["搞笑", "游戏", "美食", "科技", "动漫", "日常", "数码", "手工", "汽车", "运动"]

    @staticmethod
    def _fix_cover_url(cover_url: str) -> str:
        """Sửa lỗi CDN thiếu HTTPS khiến thumbnail không hiện"""
        if not cover_url:
            return ""
        if cover_url.startswith("//"):
            return "https:" + cover_url
        elif not cover_url.startswith("http"):
            return "https://" + cover_url
        return cover_url

    @staticmethod
    def _format_duration(duration_input: Any) -> str:
        """Format số giây thành MM:SS chuẩn YouTube"""
        if not duration_input:
            return "0:00"
        if isinstance(duration_input, str) and ":" in duration_input:
            return duration_input
        try:
            total_sec = int(duration_input)
            hrs = total_sec // 3600
            mins = (total_sec % 3600) // 60
            secs = total_sec % 60
            return f"{hrs}:{mins:02d}:{secs:02d}" if hrs > 0 else f"{mins}:{secs:02d}"
        except (ValueError, TypeError):
            return str(duration_input)

    @classmethod
    async def fetch_home_recommendations(cls, page: int = 1, count: int = 16) -> List[Dict[str, Any]]:
        """Cào video Gợi ý Trang chủ - Đảm bảo lọc sạch bvid rỗng và trả về đủ count video"""
        # Chọn từ khóa xoay vòng dựa trên số trang và random nhẹ
        topic_idx = (page - 1 + random.randint(0, 3)) % len(cls.DYNAMIC_TOPICS)
        current_topic = cls.DYNAMIC_TOPICS[topic_idx]

        # Thử cào qua Search API theo từ khóa hot trước
        results = await cls.search_videos(keyword=current_topic, page=page, count=count)
        if len(results) >= count:
            return results[:count]

        # Dự phòng gọi endpoint popular (Cào dư 24 item để trừ hao rác)
        fetch_limit = count + 8
        params = {"pn": page, "ps": fetch_limit}
        async with httpx.AsyncClient(headers=cls.HEADERS, timeout=8.0) as client:
            try:
                response = await client.get(cls.BASE_HOME_URL, params=params)
                if response.status_code == 200:
                    data = response.json()
                    raw_list = data.get("data", {}).get("list", [])
                    
                    parsed_results = []
                    for item in raw_list:
                        bvid = item.get("bvid", "").strip()
                        duration_str = cls._format_duration(item.get("duration", 0))

                        # 🚨 BỘ LỌC RÁC: Bỏ qua video thiếu BVID hoặc bài quảng cáo (0:00)
                        if not bvid or duration_str == "0:00":
                            continue

                        parsed_results.append({
                            "bvid": bvid,
                            "title": item.get("title", ""),
                            "cover": cls._fix_cover_url(item.get("pic", "")),
                            "duration": duration_str,
                            "play": item.get("stat", {}).get("view", 0),
                            "like": item.get("stat", {}).get("like", 0),
                            "author": item.get("owner", {}).get("name", "N/A"),
                            "pubdate": item.get("pubdate", 0),
                            "platform": "bilibili"
                        })
                        
                        # Đã đủ số lượng yêu cầu thì dừng ngay
                        if len(parsed_results) == count:
                            break

                    return parsed_results
            except Exception as e:
                print(f"[BILIBILI HOME CRAWL ERROR]: {e}")
        
        return []

    @classmethod
    async def search_videos(cls, keyword: str, page: int = 1, count: int = 16) -> List[Dict[str, Any]]:
        """Cào theo từ khóa tìm kiếm - Lọc rác BVID rỗng"""
        fetch_limit = count + 8
        params = {
            "search_type": "video",
            "keyword": keyword,
            "page": page,
            "page_size": fetch_limit
        }
        async with httpx.AsyncClient(headers=cls.HEADERS, timeout=8.0) as client:
            try:
                response = await client.get(cls.BASE_SEARCH_URL, params=params)
                if response.status_code != 200:
                    return []

                data = response.json()
                raw_list = data.get("data", {}).get("result", [])

                results = []
                for item in raw_list:
                    bvid = item.get("bvid", "").strip()
                    duration_str = cls._format_duration(item.get("duration", "0:00"))

                    # 🚨 BỘ LỌC RÁC: Bỏ qua video thiếu BVID hoặc quảng cáo
                    if not bvid or duration_str == "0:00":
                        continue

                    clean_title = item.get("title", "").replace('<em class="keyword">', '').replace('</em>', '')
                    
                    results.append({
                        "bvid": bvid,
                        "title": clean_title,
                        "cover": cls._fix_cover_url(item.get("pic", "")),
                        "duration": duration_str,
                        "play": item.get("play", 0),
                        "like": item.get("favorites", 0),
                        "author": item.get("author", "N/A"),
                        "pubdate": item.get("pubdate", 0),
                        "platform": "bilibili"
                    })

                    if len(results) == count:
                        break

                return results
            except Exception as e:
                print(f"[BILIBILI SEARCH ERROR]: {e}")
                return []

File: server/bilibiliCrawler.ts
export interface VideoItem {
  bvid: string;
  title: string;
  title_vi?: string;
  cover: string;
  duration: string;
  play: number;
  like: number;
  author: string;
  pubdate: number;
  platform: string;
}

export class BilibiliCrawler {
  private static readonly BASE_SEARCH_URL = "https://api.bilibili.com/x/web-interface/wbi/search/type";
  private static readonly BASE_HOME_URL = "https://api.bilibili.com/x/web-interface/popular";

  private static readonly HEADERS: Record<string, string> = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
    "Referer": "https://www.bilibili.com/"
  };

  private static readonly DYNAMIC_TOPICS = ["搞笑", "游戏", "美食", "科技", "动漫", "日常", "数码", "手工", "汽车", "运动"];

  public static fixCoverUrl(coverUrl?: string): string {
    if (!coverUrl) return "";
    if (coverUrl.startsWith("//")) return "https:" + coverUrl;
    if (!coverUrl.startsWith("http")) return "https://" + coverUrl;
    return coverUrl;
  }

  public static formatDuration(durationInput: any): string {
    if (!durationInput) return "0:00";
    if (typeof durationInput === "string" && durationInput.includes(":")) {
      return durationInput;
    }
    try {
      const totalSec = parseInt(String(durationInput), 10);
      if (isNaN(totalSec)) return String(durationInput);
      const hrs = Math.floor(totalSec / 3600);
      const mins = Math.floor((totalSec % 3600) / 60);
      const secs = totalSec % 60;
      const pad = (n: number) => n.toString().padStart(2, "0");
      return hrs > 0 ? `${hrs}:${pad(mins)}:${pad(secs)}` : `${mins}:${pad(secs)}`;
    } catch {
      return String(durationInput);
    }
  }

  public static async fetchHomeRecommendations(
    page: number = 1,
    count: number = 16,
    order: string = "totalrank",
    timeRange: string = "all"
  ): Promise<VideoItem[]> {
    // If user requested custom sorting (pubdate / click) or time filter, use topic search with parameters
    if (order !== "totalrank" || (timeRange && timeRange !== "all")) {
      const topicIdx = Math.abs(page - 1 + Math.floor(Math.random() * 4)) % this.DYNAMIC_TOPICS.length;
      const currentTopic = this.DYNAMIC_TOPICS[topicIdx];
      return await this.searchVideos(currentTopic, page, count, order, timeRange);
    }

    const topicIdx = Math.abs(page - 1 + Math.floor(Math.random() * 4)) % this.DYNAMIC_TOPICS.length;
    const currentTopic = this.DYNAMIC_TOPICS[topicIdx];

    const results = await this.searchVideos(currentTopic, page, count, order, timeRange);
    if (results.length >= count) {
      return results.slice(0, count);
    }

    const fetchLimit = count + 8;
    const url = `${this.BASE_HOME_URL}?pn=${page}&ps=${fetchLimit}`;
    try {
      const response = await fetch(url, {
        headers: this.HEADERS,
        signal: AbortSignal.timeout(8000)
      });

      if (response.ok) {
        const data = await response.json();
        const rawList = data?.data?.list || [];

        const parsedResults: VideoItem[] = [];
        for (const item of rawList) {
          const bvid = (item.bvid || "").trim();
          const durationStr = this.formatDuration(item.duration);

          if (!bvid || durationStr === "0:00") {
            continue;
          }

          parsedResults.push({
            bvid,
            title: item.title || "",
            cover: this.fixCoverUrl(item.pic || ""),
            duration: durationStr,
            play: item.stat?.view ?? 0,
            like: item.stat?.like ?? 0,
            author: item.owner?.name || "N/A",
            pubdate: item.pubdate ?? 0,
            platform: "bilibili"
          });

          if (parsedResults.length === count) {
            break;
          }
        }

        return parsedResults;
      }
    } catch (e) {
      console.error("[BILIBILI HOME CRAWL ERROR]:", e);
    }

    return results;
  }

  public static async searchVideos(
    keyword: string,
    page: number = 1,
    count: number = 16,
    order: string = "totalrank",
    timeRange: string = "all"
  ): Promise<VideoItem[]> {
    const isTimeFiltered = timeRange && timeRange !== "all";
    // When time filtering is requested, request a larger page size so we still have plenty of results
    const fetchLimit = isTimeFiltered ? 45 : Math.min(50, count + 8);

    // Normalize order parameter for Bilibili
    let bilibiliOrder = "totalrank";
    if (order === "pubdate") bilibiliOrder = "pubdate";
    else if (order === "click") bilibiliOrder = "click";
    else if (order === "stow") bilibiliOrder = "stow";

    // If timeRange is set (e.g. today/week) but order is default totalrank,
    // sorting by pubdate gives more recent videos within that range
    if (isTimeFiltered && order === "totalrank") {
      bilibiliOrder = "pubdate";
    }

    const url = `${this.BASE_SEARCH_URL}?search_type=video&keyword=${encodeURIComponent(keyword)}&page=${page}&page_size=${fetchLimit}&order=${bilibiliOrder}`;

    try {
      const response = await fetch(url, {
        headers: this.HEADERS,
        signal: AbortSignal.timeout(8000)
      });

      if (!response.ok) {
        return [];
      }

      const data = await response.json();
      const rawList = data?.data?.result || [];

      // Calculate min timestamp for timeRange in seconds
      const nowSec = Math.floor(Date.now() / 1000);
      let minPubdateSec = 0;
      if (timeRange === "today") {
        minPubdateSec = nowSec - 24 * 60 * 60; // past 24 hours
      } else if (timeRange === "week") {
        minPubdateSec = nowSec - 7 * 24 * 60 * 60; // past 7 days
      } else if (timeRange === "month") {
        minPubdateSec = nowSec - 30 * 24 * 60 * 60; // past 30 days
      } else if (timeRange === "year") {
        minPubdateSec = nowSec - 365 * 24 * 60 * 60; // past 365 days
      }

      const results: VideoItem[] = [];
      for (const item of rawList) {
        const bvid = (item.bvid || "").trim();
        const durationStr = this.formatDuration(item.duration);

        if (!bvid || durationStr === "0:00") {
          continue;
        }

        const pubdate = item.pubdate || item.senddate || 0;

        // Apply time range filter if specified
        if (minPubdateSec > 0 && pubdate > 0 && pubdate < minPubdateSec) {
          continue;
        }

        const cleanTitle = (item.title || "")
          .replace(/<em class="keyword">/g, "")
          .replace(/<\/em>/g, "");

        results.push({
          bvid,
          title: cleanTitle,
          cover: this.fixCoverUrl(item.pic || ""),
          duration: durationStr,
          play: item.play ?? 0,
          like: item.favorites ?? 0,
          author: item.author || "N/A",
          pubdate,
          platform: "bilibili"
        });

        if (results.length === count) {
          break;
        }
      }

      // If we filtered by time, sort results according to requested order
      if (order === "click") {
        results.sort((a, b) => (b.play || 0) - (a.play || 0));
      } else if (order === "pubdate") {
        results.sort((a, b) => (b.pubdate || 0) - (a.pubdate || 0));
      }

      return results;
    } catch (e) {
      console.error("[BILIBILI SEARCH ERROR]:", e);
      return [];
    }
  }
}
